File: SecondYearCS/OOP/CivitasRB/lib/TemplateCasilla.rb
require_relative 'sorpresa.rb'
require_relative 'tipo_sorpresa.rb'
require_relative 'titulo_propiedad.rb'
require_relative 'mazo_sorpresa.rb'
require_relative 'tipo_casilla.rb'
require_relative 'diario.rb'
require_relative 'dado.rb'
module CivitasRB
  class Casilla
    attr_accessor :nombre, :tituloPropiedad
    def initialize(args)
      args = defaults.merge(args)
      @nombre = args[:nombre]
      @importe = args[:importe]
      @carcel = args[:carcel]
      @sorpresa = args[:sorpresa]
      @tituloPropiedad = args[:tituloPropiedad]
      @mazo = args[:mazo]
      @tipo = args[:tipo]
      if @tituloPropiedad != nil
        @nombre = @tituloPropiedad.nombre
        @tipo = TipoCasilla::CALLE
      end
      if @importe != 0
        @tipo = TipoCasilla::IMPUESTO        
      end
      if @carcel != 0
        @tipo = TipoCasilla::JUEZ
      end
      if @mazo != nil
        @tipo = TipoCasilla::SORPRESA
      end
      if @tipo == nil
        @tipo = TipoCasilla::DESCANSO
      end
    end
    
    def defaults
      {:nombre => "", :importe => 0, :carcel => 7,
        :mazo => nil, :sorpresa => nil, :tipo => nil, :tituloPropiedad => nil
      }
    end
    
    def jugadorCorrecto(act,todos)
      act >=0 && act < todos.length
    end
    
    def to_s
     res = "\nCasilla: " + @nombre
     if(@tituloPropiedad != nil)
       res += " con: " + "\n 1- precioCompra: " + @tituloPropiedad.precioCompra.to_s +
         "\n 2- precioEdificación: " + @tituloPropiedad.precioEdificar.to_s +
         "\n 3- AlquilerBase: " + @tituloPropiedad.alquilerBase.to_s
     end
     return res
    end
    
    #Informar al diario de la ocurrencia 
    def informe(act,todos)
      if self.jugadorCorrecto(act,todos)
        Diario.instance.ocurre_evento("Jugador: " + 
                    todos.at(act).nombre + self.to_s)
      end
    end
    
    #LLamar al metodo paga impuesto de la clase jugador
    def recibeJugador_impuesto(act,todos)
      if self.jugadorCorrecto(act,todos)
        self.informe(act,todos)
        todos.at(act).pagaImpuesto(@importe)
      end
    end
    
    #Encarcelar al jugador 
    #Pre-Condicion, asegurar que la instancia esta bien
    #inicializada en la clase tablero. (sorpresa!=nil)
    def recibeJugador_juez(act,todos)
      if self.jugadorCorrecto(act,todos)
        self.informe(act,todos)
        todos.at(act).encarcelar(@carcel)
      end
    end
    
    def recibeJugador_calle(act,todos)
      if self.jugadorCorrecto(act, todos)
        self.informe(act,todos)
        jugadorActual = todos.at(act)
        if !@tituloPropiedad.tienePropietario
          if jugadorActual.puedeComprarCasilla
            @tituloPropiedad.tramitarAlquiler(jugadorActual)
          end
        end
      end
    end
    
    def recibeJugador_sorpresa(act,todos)
      if self.jugadorCorrecto(act, todos)
        self.informe(act,todos)
        sorpresa = @mazo.siguiente()
        sorpresa.aplicarAJugador(act,todos)
      end
    end
    
    def recibeJugador(act,todos)
      case @tipo
      when TipoCasilla::CALLE
        recibeJugador_calle(act,todos)
      when TipoCasilla::IMPUESTO
        recibeJugador_impuesto(act,todos)
      when TipoCasilla::JUEZ
        recibeJugador_juez(act,todos)
      when TipoCasilla::SORPRESA
        recibeJugador_sorpresa(act,todos)
      else
        self.informe(act,todos)
      end
    end

    def getTituloPropiedad
      return @tituloPropiedad
    end
  end
end

File: SecondYearCS/OOP/OldStuff/TemplateCasilla.java
package civitas;
import java.util.ArrayList;

public class TemplateCasilla {

    //atributos de clase
    private String nombre;
    private int carcel;
    private float importe;

    //atributos de referencia
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad; // tipo:: CALLE
    private Sorpresa sorpresa;               // tipo:: SORPRESA
    private MazoSorpresa mazo;              // tipo:: SORPRESA
    
    void init(){
        tipo = null;
        tituloPropiedad = null;
        sorpresa = null;
        mazo = null;
        carcel = -1;
        importe = 0;
    }
    
    TemplateCasilla ( String nombre ) {
        init();
        this.nombre = nombre;
        this.tipo = TipoCasilla.DESCANSO;
    }
    //Tipo Calle
    TemplateCasilla(TituloPropiedad titulo){
        init();
        tipo = TipoCasilla.CALLE;
        this.tituloPropiedad = titulo;    
        this.nombre = titulo.getNombre();
    }
    //Tipo Impuesto
    TemplateCasilla(float cantidad, String nombre){
        init();
        this.importe = cantidad;
        this.nombre = nombre;
        this.tipo = TipoCasilla.IMPUESTO;
    }
    //Tipo Juez
    TemplateCasilla(int numCasillaCarcel, String nombre){
        init();
        this.carcel = numCasillaCarcel;
        this.nombre = nombre;
        this.tipo = TipoCasilla.JUEZ;
    }
    //Tipo Sorpresa
    TemplateCasilla(MazoSorpresa mazo, String nombre){
        init();
        this.nombre = nombre;
        this.mazo = mazo;
        this.tipo = TipoCasilla.SORPRESA;
    }

    boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos){
        boolean res = false;
        if( actual >= 0 && todos.size() > actual ){
            res = true;
        }
        return res;
    }
    
    void informe(int actual, ArrayList<Jugador> todos){
        if( this.jugadorCorrecto(actual, todos) ){
            Diario.getInstance().ocurreEvento( "Jugador: " +
                     todos.get(actual).getNombre() + "\n " +  this.toString() );
        }
    }
    
    void recibeJugador_impuesto(int actual, ArrayList<Jugador> todos){
        if( this.jugadorCorrecto(actual, todos) ){
            this.informe(actual,todos);
            todos.get(actual).pagaImpuesto(this.importe);
        }
    }

    void recibeJugador_juez(int actual, ArrayList<Jugador> todos){
        if( this.jugadorCorrecto(actual, todos) ){
            this.informe(actual,todos);
            todos.get(actual).encarcelar(this.carcel);
        }
    }
    
    void recibeJugador_calle (int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual,todos)){
            this.informe(actual,todos);
            Jugador jugadorActual = todos.get(actual);
            if(!this.tituloPropiedad.tienePropietario()){
                if(jugadorActual.puedeComprarCasilla())
                    this.tituloPropiedad.tramitarAlquiler(jugadorActual);
            }
        }
    }
    void recibeJugador_sorpresa (int actual, ArrayList<Jugador> todos) throws Exception{
        if(this.jugadorCorrecto(actual,todos)){
            this.informe(actual,todos);
            Sorpresa sorpresa = this.mazo.siguiente();
            sorpresa.aplicarAJugador(actual, todos);
        }
    }
    
    void recibeJugador(int actual, ArrayList<Jugador>todos) throws Exception{
        switch(this.tipo){
            case CALLE:
                recibeJugador_calle(actual,todos);
                break;
            case IMPUESTO:
                recibeJugador_impuesto(actual,todos);
                break;
            case JUEZ:
                recibeJugador_juez(actual,todos);
                break;
            case SORPRESA:
                recibeJugador_sorpresa(actual,todos);
                break;
            default:
                informe(actual,todos);  
                break;
            
        }
    }

    @Override
    public String toString(){
       String result = "Casilla: " + this.nombre;
       if(tituloPropiedad!=null){
           result += "\n 1- precio de compra: " + this.tituloPropiedad.getPrecioCompra()
               + "\n 2- precio de edificación: " + this.tituloPropiedad.getPrecioEdificar()
               + "\n 3- alquiler base: " + this.tituloPropiedad.getAlquilerBase();     
       }
       return result;
    }

    String get_nombre() {
        return nombre;
    }

    public TituloPropiedad getTituloPropiedad(){
        return this.tituloPropiedad;
    }  
}
